cli/js/compiler.ts

```diff
@@ -31,28 +32,47 @@ enum MediaType {
   Unknown = 5
 }
 
+// Warning! The values in this enum are duplicated in cli/msg.rs
+// Update carefully!
+enum CompilerReqType {
+  Compile = 1,
+  Bundle = 2,
+  Doc = 3,
+  AST = 4
+}
+
 // Startup boilerplate. This is necessary because the compiler has its own
 // snapshot. (It would be great if we could remove these things or centralize
 // them somewhere else.)
 const console = new Console(core.print);
-window.console = console;
-window.workerMain = workerMain;
+self.console = console;
+self.workerMain = workerMain;
 function denoMain(): void {
   os.start(true, "TS");
 }
-window["denoMain"] = denoMain;
+self["denoMain"] = denoMain;
 
 const ASSETS = "$asset$";
 const OUT_DIR = "$deno$";
 
 /** The format of the work message payload coming from the privileged side */
-interface CompilerReq {
+type CompilerReq = {
   rootNames: string[];
-  bundle?: string;
   // TODO(ry) add compiler config to this interface.
   // options: ts.CompilerOptions;
   configPath?: string;
   config?: string;
+} & (
+  | {
+      type: CompilerReqType.Bundle;
+      outfile: string;
+    }
+  | {
+      type: CompilerReqType.Compile | CompilerReqType.Doc | CompilerReqType.AST;
+    });
```

Comment:
I don't really understand what the & and | are doing? Is this setting defaults?

Reply:
It is a "tagged union type" where it is a dynamic type that changes it "shape" based on what the `type` tag's value is.